Mathlib/ModelTheory/PartialEquiv.lean

```diff
@@ -269,6 +273,187 @@ theorem toPartialEquiv_toEmbedding {f :  M ≃ₚ[L] N} (h : f.dom = ⊤) :
+    ).comp
+      ((DirectLimit.equiv_lift L ι (fun i ↦ (S i).dom)
+        (fun _ _ hij ↦ Substructure.inclusion (dom_le_dom (S.monotone hij)))
+        (fun i ↦ (S i).cod)
+        (fun _ _ hij ↦ Substructure.inclusion (cod_le_cod (S.monotone hij)))
+        (fun i ↦ (S i).toEquiv)
+        (fun _ _ hij _ ↦ toEquiv_inclusion_apply (S.monotone hij) _)
+      ).comp
+        (Equiv_iSup {
+          toFun := (fun i ↦ (S i).dom)
+          monotone' := monotone_dom.comp S.monotone}).symm)
+
+@[simp]
+theorem partialEquiv_limit_dom : (partialEquivLimit S).dom = iSup (fun x ↦ (S x).dom) := rfl
+
+@[simp]
+theorem partialEquiv_limit_cod : (partialEquivLimit S).cod = iSup (fun x ↦ (S x).cod) := rfl
+
+@[simp]
+lemma partialEquivLimit_comp_inclusion {i : ι} :
+    (partialEquivLimit S).toEquiv.toEmbedding.comp (Substructure.inclusion (le_iSup _ i)) =
+    (Substructure.inclusion (le_iSup _ i)).comp (S i).toEquiv.toEmbedding := by
+  simp only [partialEquivLimit, Equiv.comp_toEmbedding, Embedding.comp_assoc]
+  rw [Equiv_isup_symm_inclusion]
+  congr
+
+theorem le_partialEquivLimit : ∀ i, S i ≤ partialEquivLimit S :=
+  fun i => ⟨le_iSup (f := fun i ↦ (S i).dom) _, by
+    simp only [partialEquiv_limit_cod, partialEquiv_limit_dom, partialEquivLimit_comp_inclusion, ←
+      Embedding.comp_assoc, subtype_comp_inclusion]⟩
+
+end DirectLimit
+
+section FGEquiv
+
+open PartialEquiv Set DirectLimit
+
+variable (M) (N) (L)
+
+/-- The type of equivalences between finitely generated substructures. -/
+abbrev FGEquiv := {f : M ≃ₚ[L] N // f.dom.FG}
+
+/-- Two structures `M` and `N` form an extension pair if the domain of any finitely-generated map
+from `M` to `N` can be extended to include any element of `M`. -/
+def extensionPair : Prop := ∀ (f : L.FGEquiv M N) (m : M), ∃ g, m ∈ g.1.dom ∧ f ≤ g
+
+variable {M N L}
+
+theorem Substructure.countable_self_FGEquiv_of_countable [Countable M] :
+    Countable { f : M ≃ₚ[L] M // f.dom.FG } := by
+  let g : { f : M ≃ₚ[L] M // f.dom.FG } →
+      Σ U : { S : L.Substructure M // S.FG }, U.val →[L] M :=
+    fun f ↦ ⟨⟨f.val.dom, f.prop⟩, (subtype _).toHom.comp f.val.toEquiv.toHom⟩
+  have g_inj : Function.Injective g := by
+    intro f f' h
+    ext
+    let ⟨⟨dom_f, cod_f, equiv_f⟩, f_fin⟩ := f
+    cases congr_arg (·.1) h
+    apply PartialEquiv.ext (by rfl)
+    simp only [g, Sigma.mk.inj_iff, heq_eq_eq, true_and] at h
+    exact fun x hx ↦ congr_fun (congr_arg (↑) h) ⟨x, hx⟩
+  have : ∀ U : { S : L.Substructure M // S.FG }, Structure.FG L U.val :=
+    fun U ↦ (U.val.fg_iff_structure_fg.1 U.prop)
+  exact Function.Embedding.countable ⟨g, g_inj⟩
+
+instance inhabited_self_FGEquiv : Inhabited (L.FGEquiv M M) :=
+  ⟨⟨⟨⊥, ⊥, Equiv.refl L (⊥ : L.Substructure M)⟩, fg_bot⟩⟩
+
+/-- Maps to the symmetric finitely-generated partial equivalence. -/
+def FGEquiv.symm (f : L.FGEquiv M N) : L.FGEquiv N M := ⟨f.1.symm, f.1.fg_iff.1 f.2⟩
+
+lemma extensionPair_symm : L.extensionPair M N ↔
```

Comment:
```suggestion
lemma extensionPair_iff_cod : L.extensionPair M N ↔
```
as `_symm` (or rather `_comm` would rather relate `L.extensionPair M N` and `L.extensionPair N M`